### ISSUE TITLE
calibrate_reads.R giving error

### ISSUE THREAD

**Issue by kmangano** [AlexanderBartholomaeus/MiMB_ribosome_profiling#1]
I have run the pipeline starting with my own BAM files.  Now I get the error pasted below while running calibrate_reads.R:

input:  
$ Rscript calibrate_reads.R ../BAMs/ ../out/calibration/calibration_3prime_config.csv  ../out/

output:
Loading required package: iterators
Loading required package: parallel
4 / 4 found in calibration config file!
Calibrating reads by length and determined nucleotide position ... Error in { : task 1 failed - "undefined columns selected"
Calls: %dopar% -> <Anonymous>
Execution halted

Thanks

**Comment by AlexanderBartholomaeus** [comment]
Dear kmangano,

did you run the verify_dependencies.R successfully? Looks like something is not working with samtool. Do you have samtools installed?

Best regards,
Alex

**Comment by kmangano** [comment]
Dear Alex,
verify_dependencies.R did run successfully.  I am using samtools V 1.10 with htslib 1.10.2
Thanks,
Kyle

**Comment by AlexanderBartholomaeus** [comment]
Can you share your calibration_3prime_config.csv?

After I read the error again "undefined columns selected" I think this indicates that it cannot refer to the correct column.

**Comment by kmangano** [comment]
Attaching it here.

**Comment by kmangano** [comment]
[calibration_3prime_config.zip](https://github.com/AlexanderBartholomaeus/MiMB_ribosome_profiling/files/5203966/calibration_3prime_config.zip)

**Comment by AlexanderBartholomaeus** [comment]
Can you copy the attached file to the folder with scripts, where calibrate_reads.R is and execute the command again with a changed debug script, please. Command below: just add ..._debug to the script name before .R :

Rscript calibrate_reads_debug.R ../BAMs/ ../out/calibration/calibration_3prime_config.csv ../out/

I added a lot of debugging messages, so let's see what is says.

[calibrate_reads_debug.zip](https://github.com/AlexanderBartholomaeus/MiMB_ribosome_profiling/files/5204082/calibrate_reads_debug.zip)

**Comment by kmangano** [comment]
Actually it gave the same error message. 
Calibrating reads by length and determined nucleotide position ... Error in { : task 1 failed - "undefined columns selected"
Calls: %dopar% -> 'Anonymous'

the 'Anonymous' was not included in the original comment due to <> being instead of ''

**Comment by kmangano** [comment]
I am not experienced in R, but maybe this will help.  I removed the %doPar%  below is the error from the debugging script:

Error in print(cali_conf[, 1] == par_read_length[i]) :
  object 'i' not found
Execution halted

**Comment by AlexanderBartholomaeus** [comment]
Yes maybe dopar is the problem. Can you replace the line 87 (in the changed script) with the %dopar% command with the following:

```
for(i in 1:length(par_files)){ 

```

and re-run again please

**Comment by kmangano** [comment]
Attached is the updated script and here is the output:

Loading required package: iterators
Loading required package: parallel
4 / 4 found in calibration config file!
Calibrating reads by length and determined nucleotide position ... [1] "Debug i"
 [1] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[1] NA NA NA NA NA
Error in `[.data.frame`(cali_conf, cali_conf[, 1] == par_read_length[i],  :
  undefined columns selected
Calls: [ -> [.data.frame
Execution halted
[calibrate_reads_debug_no_doPar.zip](https://github.com/AlexanderBartholomaeus/MiMB_ribosome_profiling/files/5219183/calibrate_reads_debug_no_doPar.zip)

**Comment by AlexanderBartholomaeus** [comment]
I thought that the script checks possibly missing files, however it seems that the requested files are not present. It would be really nice if you could execute again. Thank you!
[calibrate_reads_debug_20200914.zip](https://github.com/AlexanderBartholomaeus/MiMB_ribosome_profiling/files/5220054/calibrate_reads_debug_20200914.zip)

**Comment by kmangano** [comment]
Happy to keep executing scripts!  Ultimately the same error again:

Loading required package: iterators
Loading required package: parallel
[1] "inFiles: 24915_Ctrl.sort.bam" "inFiles: 24915_Sar.sort.bam"
[3] "inFiles: 24915_Tet.sort.bam"  "inFiles: 24915_Tig.sort.bam"
   length X24915_Ctrl X24915_Sar X24915_Tet X24915_Tig
1      24          14         14         14         14
2      25          14         14         14         15
3      26          14         14         14         15
4      27          14         15         15         15
5      28          14         15         15         15
6      29          14         14         15         15
7      30          14         15         15         14
8      31          15         15         15         15
9      32          15         15         15         15
10     33          15         15         15         15
11     34          15         15         15         15
12     35          15         15         15         15
13     36          15         15         15         15
14     37          15         15         15         15
15     38          15         15         15         15
[1] "in_f: 24915_Ctrl" "in_f: 24915_Sar"  "in_f: 24915_Tet"  "in_f: 24915_Tig"
[1] "in_f2: "
4 / 4 found in calibration config file!
[1] "par_files: "
[1] "par_read_length: "
Calibrating reads by length and determined nucleotide position ... [1] "Debug 1"
 [1] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[1] NA NA NA NA NA
Error in `[.data.frame`(cali_conf, cali_conf[, 1] == par_read_length[i],  :
  undefined columns selected
Calls: [ -> [.data.frame
Execution halted

**Comment by AlexanderBartholomaeus** [comment]
Yes same error, but now I know the error! It is the "stupid" R definition of column names, which does not allow numbers and adds upper X. I will fix this and post a new script which should then hopefully run!

**Comment by kmangano** [comment]
Great!  You know this because the object "in_f2" has no value?  

I was getting worried that this was because I started with my own sorted BAM files, instead of those made with your scripts.

**Comment by AlexanderBartholomaeus** [comment]
Yes exactly because this is empty. Additionally there is a bug in the output below which should show 0/4 found. 

Attached a script that should work (unless you have bam files starting with X in the name). It still contains a lot of debugging but this messages. I will write a proper function for the main script later ... need to find a more secure way handeling all possible filenames. 

How did you came across the tool? Since the article is not out yet?

[calibrate_reads_debug_20200914_2.zip](https://github.com/AlexanderBartholomaeus/MiMB_ribosome_profiling/files/5220528/calibrate_reads_debug_20200914_2.zip)

**Comment by kmangano** [comment]
program looks to be running successfully, but looks like it will take a long time.  Thanks!

I am a graduate student in the Mankin lab who didn't have any informatics experience.  I came to the Ignatova lab in Hamburg  to learn some basics of NGS analysis tools 2 years ago.  Baban and Zoya shared this Github link with me.

**Comment by kmangano** [comment]
The script generated calibrated BAM files named according to my sample IDs.  However, I ran into a problem running merge_reads.R "no BAM files found"

Check lines 35 and 36 of the attached script to see my workaround with the filename pattern line reading in BAMs.  My modded script generated merged BAMs without throwing errors, but now those BAMs are giving errors in the coverage_start_stop.R script. 

Error: Sorted input specified, but the file ../Calibrate_results/calibration/calibrated/24915_Ctrl_30_calibrated.bam has the following out of order record

I think I can move forward with calibrated BAMs to use in with custom bed files in coverageBed, I just wanted to let you know.

Thanks for the help!
[merge_reads.zip](https://github.com/AlexanderBartholomaeus/MiMB_ribosome_profiling/files/5221385/merge_reads.zip)

**Comment by AlexanderBartholomaeus** [comment]
I think my plan was to generate the coverage plot only on the file with merge calibrated read of different length. However, from the code it looks like all bam files in the input folder are taken. I will check this for the next update.

Did you get the coverage plots for the merged calibrated reads? Otherwise just copy the file to an empty folder and give this folder as input.

Thank you for your patience!

**Comment by kmangano** [comment]
Ah I see, good point to move the calibrated, merged bams to a new location. 

For the next update, there is one more problem I encountered when running the series of scripts.  Any time bedtools coverage was called within an R script, I added the parameter -sorted.  Without that argument, I was getting memory usage errors because my BAM file was too large.

**Comment by kmangano** [comment]
Hi Alex,
I have one more question about the coverage_start_stop.R metagene plots generated before and after calibration.  From the attached PDFs, you can see a huge difference before and after calibration.  This is a bigger difference than I expected, because a 15 nt offset was used for 12/15 read lengths.  Is the pre-calibration metagene plot essentially just using the 3' ends of the reads?  In that case I would expect to see a more sharp peak at the +15 position from the start codon.  Do you have an explanation of why the pre-calibrated metagene is so broad or smoothed?

Thanks

[coverage_24915_Tet.sort.pdf](https://github.com/AlexanderBartholomaeus/MiMB_ribosome_profiling/files/5225914/coverage_24915_Tet.sort.pdf)
[coverage_24915_Tet_sorted_calibrated.pdf](https://github.com/AlexanderBartholomaeus/MiMB_ribosome_profiling/files/5225920/coverage_24915_Tet_sorted_calibrated.pdf)

**Comment by AlexanderBartholomaeus** [comment]
Oh interesting point with the coverageBed, I never used -sorted because I always had 64GB of RAM available and no problem.

Regarding your plot: One plot is done using the full reads, i.e. each nucleotide of a read is counted and thus the curve get smoothed. Additionally, you might also use all read length which will also smooth because certain read length are more specific than others. The plot with the sharp peak is generated using only one nucleotide of each read (mid nucleotide of P-site codon).
